### PR TITLE
Feature/727 target rates for tokensale

### PR DESCRIPTION
**Pull request by gorgos**
https://starbase.atlassian.net/browse/SP-729

### CODE REVIEW

File: contracts/TokenSale.sol [hunk 1]
```diff
@@ -334,6 +318,71 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
...
+        require(
+            targetRatesTimestamps[0] == startTime,
+            'First target rate timestamp should match startTime!'
+        );
+
+        for (uint256 i = 0; i < targetRates.length; i++) {
+            if (i > 0) {
+                require(
+                    targetRatesTimestamps[i-1] < targetRatesTimestamps[i],
+                    'Target rates timestamps should be ordered!'
```
Comment (franzejr) [inline contracts/TokenSale.sol]: sorted?

File: contracts/TokenSale.sol [hunk 2]
```diff
@@ -334,6 +318,71 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
...
+            }
+
+            if (i == targetRates.length - 1) {
+               require(
+                    targetRatesTimestamps[i] < endTime,
+                    'All target rate timestamps should be before endTime!'
+                ); 
+            }
+
+            require(targetRates[i] > 0, 'All target rates must above 0!');
```
Comment (franzejr) [inline contracts/TokenSale.sol]: must be*

File: contracts/TokenSale.sol [hunk 3]
```diff
@@ -334,6 +318,71 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
...
+            targetRatesTimestamps.length - 1
+        );
+    }
+
+    /**
+     * @dev Check for new valid rate and update. Automatically called when
+     * purchasing tokens. 
+     */
+    function checkForNewRateAndUpdate() public {
+        (, uint256 targetRateIndex) = getCurrentRate();
```
Comment (franzejr) [inline contracts/TokenSale.sol]: this `(,` is normal here?

File: contracts/TokenSale.sol [hunk 4]
```diff
@@ -226,12 +210,12 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
         uint256 weiRefund;
 
         // calculate token amount to be created
-        uint256 tokens = weiAmount.mul(rate);
+        uint256 tokens = weiAmount.mul(targetRates[currentTargetRateIndex]);
```
[the remark this answers is not in the document]
Comment (gorgos) [reply contracts/TokenSale.sol]: 👍

File: contracts/TokenSale.sol [hunk 5]
```diff
@@ -334,6 +318,71 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
         if (investedStarRefund > 0) starToken.transfer(msg.sender, investedStarRefund);
     }
 
+    function verifyTargetRates() internal view {
+        require(
+            targetRates.length == targetRatesTimestamps.length,
+            'Target rates and target rates timestamps should match!'
```
[the remark this answers is not in the document]
Comment (gorgos) [reply contracts/TokenSale.sol]: Length, thank you 👍

File: contracts/TokenSale.sol (hunk 1 above)
Comment (gorgos) [reply contracts/TokenSale.sol]: 👍

File: contracts/TokenSale.sol (hunk 3 above)
Comment (gorgos) [reply contracts/TokenSale.sol]: Yeah, ignoring the first return value as `getCurrentRate` returns two arguments, but we only need the second one.

File: contracts/StarEthRate.sol [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            'Please pass a decimalCorrectionFactor above 0!'
+        );
+        require(_initialStarEthRate > 0, 'Please pass a starEthRate above 0!');
+
+        decimalCorrectionFactor = _decimalCorrectionFactor;
+        starEthRate = _initialStarEthRate;
+    }
+
+    function setDecimalCorrectionFactor(
+        uint256 _newdecimalCorrectionFactor
```
Comment (gustavoguimaraes) [inline contracts/StarEthRate.sol]: small aesthetic change: `_newdecimalCorrectionFactor` to `_newDecimalCorrectionFactor`

File: contracts/TokenSale.sol [hunk 7]
```diff
@@ -1,23 +1,31 @@
...
+
 import "./FundsSplitterInterface.sol";
+import "./StarEthRateInterface.sol";
+import "./TokenSaleInterface.sol";
+import "./Whitelist.sol";
 
 /**
  * @title Token Sale contract - crowdsale of company tokens.
  * @author Gustavo Guimaraes - <gustavo@starbase.co>
+ * @author Markus Waas - <markus@starbase.co>
```
Comment (gustavoguimaraes) [inline contracts/TokenSale.sol]: 👍

File: contracts/TokenSale.sol [hunk 8]
```diff
@@ -43,71 +52,69 @@ contract TokenSale is FinalizableCrowdsale, Pausable {
...
      * @param _startTime The timestamp of the beginning of the crowdsale
      * @param _endTime Timestamp when the crowdsale will finish
-     * @param _whitelist contract containing the whitelisted addresses
-     * @param _starToken STAR token contract address
-     * @param _companyToken ERC20 contract address that has minting capabilities
-     * @param _tokenOwnerAfterSale Address that this TokenSale will pass the token ownership to after it's finished. Only works when TokenSale mints tokens, otherwise must be `0x0`.
-     * @param _rate The token rate per ETH
-     * @param _starRatePer1000 The token rate per 1/1000 STAR
-     * @param _wallet FundsSplitter wallet that redirects funds to client and Starbase.
+     * @param _externalAddresses contract containing the whitelisted addresses
```
Comment (gustavoguimaraes) [inline contracts/TokenSale.sol]: Fascinating. Good idea. Never seen it done like this. Well done

Comment (gorgos) [reply contracts/TokenSale.sol]: Whops. The comment wasnt changed.

**Comment by gorgos** [comment]
> Looks good to me. These are major changes. Please ensure you have tests for all edge cases especially for verifying rates and their corresponding timeStamps.

@gustavoguimaraes REVIEW(https://github.com/StarbaseCo/starbase-platform-smart-contracts/pull/47/commits/3ed95f8cceeedf7bc9c853afa5d3f08ca3f1bd52)

Added some more edge case tests. Anything missing in your opinion?

**Comment by gustavoguimaraes** [comment]
Should be okay
